cuda/solver/common_trs_kernels.cuh

```diff
@@ -492,8 +506,10 @@ void sptrsv_naive_caching(std::shared_ptr<const CudaExecutor> exec,
     x->fill(nan<ValueType>());
 
     Array<bool> nan_produced(exec, {false});
+    Array<IndexType> atomic_counter(exec, {IndexType{}});
```

Comment:
Comment: For small matrices, both allocating and copying the data over takes a long time.
For initialization, I recommend using a initialize kernel, which is significantly faster (I don't have the numbers anymore, but it made quite a big difference using a kernel compared to copying the data).

Reply:
Are you talking about the marked section? This is just two single values, did you mean those with "small matrices"?

Reply:
Yes, I am talking about the marked section.
Both boil down to one `cudaMalloc` and one `cudaMemcpy` each (since you provide an initial value, this will be copied through `cudaMemcpy`).
If you remove the initialization (making it `Array<type> val(exec)`), you don't have the `cudaMemcpy`.
Having a kernel that initializes the data instead of copying is faster than copying single values, so only allocating (no initialization in the `gko::Array`) followed by an initialization kernel is faster than the current method. However, that is only noticeable for small matrices, so I wouldn't consider important. I just wanted to mention it.

Reply:
Now I see what you mean. Sure, I'll include that in the review changes, which I will push probably tomorrow.